haiku/_src/integration/bfloat16_test.py

```diff
@@ -54,23 +54,35 @@ def g(x):
     x = jax.random.uniform(rng, shape)
     params, state = init_fn(rng, x)
 
-    # Cast f32 to bf16.
-    f32_to_bf16 = (
-        lambda v: v.astype(jnp.bfloat16) if v.dtype == jnp.float32 else v)
-    params, state = jax.tree_map(f32_to_bf16, (params, state))
+    # Cast f32 to test_dtype.
+    f32_to_test_dtype = (
+        lambda v: v.astype(test_dtype) if v.dtype == jnp.float32 else v)
+    params, state = jax.tree_map(f32_to_test_dtype, (params, state))
 
-    # bf16 in should result in bf16 out.
-    x = x.astype(jnp.bfloat16)
+    # test_dtype in should result in test_dtype out.
+    x = x.astype(test_dtype)
 
     for _ in range(2):
       y, state = apply_fn(params, state, rng, x)
 
-      def assert_bf16(v):
+      def assert_dtype(v):
         if v.dtype != jnp.int32:
-          self.assertEqual(v.dtype, jnp.bfloat16)
+          self.assertEqual(v.dtype, test_dtype)
+
+      jax.tree_map(assert_dtype, y)
+      jax.tree_map(assert_dtype, state)
+
+
+class Bfloat16Test(DTypeTestCase):
```

Comment:
I really like how you've made this generic over dtype, I think we should go further.

I think it's worth renaming this file to `dtype_test.py` and adding cases for bf16, f64 in here (rather than having one file per dtype). The one tricky part is that we'll need to have a runtime check for which device we are running on and skip tese as appropriate (e.g. bf16 only on tpu, f64 only on cpu).

I think you can do that with something like this:

```python
if jax.local_devices()[0].platform not in supported_platforms:
  self.skipTest('dtype not supported')
```

I think the values for platform are `cpu` `gpu` and `tpu`.

Reply:
I used a separate file because jax documentation [says](https://jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html#Double-(64bit)-precision) that float64 configuration should be done at startup. Won't that cause issues if testing for both dtypes is done in the same file?

Reply:
Thank you for pointing that out, you're right that this should be separated out.